README.md
# bs-game-of-life

```
yarn
yarn build
yarn start
```

Go to `localhost:8080`


ChangeLog.md
# Revision history for bs-game-of-life

## 2018.6.24

* Implement an HTML5 canvas with the rules of Conway's Game of Life.
* Click on a cell to toggle if it is dead or alive.
* Reset the board to a random state.
* Step button to run make a single tick.
* Play/Pause button to run or pause the game.
* Select option for speed of game.


src/GameOfLife.ml
module Element = struct
  type t
  external setInnerHTML : t -> string -> unit = "innerHTML" [@@bs.set]

  external offsetLeft : t -> int = "" [@@bs.get]
  external offsetTop : t -> int = "" [@@bs.get]
  external scrollLeft : t -> int = "" [@@bs.get]
  external scrollTop : t -> int = "" [@@bs.get]
  external offsetParent : t -> t option = "" [@@bs.get] [@@bs.return null_to_opt]
end

module Context = struct
  type t
  external setFillStyle : t -> string -> unit = "fillStyle" [@@bs.set]
  external setStrokeStyle : t -> string -> unit = "strokeStyle" [@@bs.set]
  external setLineWidth : t -> float -> unit = "lineWidth" [@@bs.set]

  external fillRect : int -> int -> int -> int -> unit = "" [@@bs.send.pipe: t]
  external moveTo : float -> float -> unit = "" [@@bs.send.pipe: t]
  external lineTo : float -> float -> unit = "" [@@bs.send.pipe: t]
  external beginPath : unit = "" [@@bs.send.pipe: t]
  external closePath : unit = "" [@@bs.send.pipe: t]
  external stroke : unit = "" [@@bs.send.pipe: t]
end

module Window = struct
  external requestAnimationFrame: (unit -> unit) -> unit = "" [@@bs.val]
end

module Mouse = struct
  type t
  external getElementById : string -> t option = "document.getElementById" [@@bs.val] [@@bs.return null_to_opt]
  external setInnerHTML : t -> string -> unit = "innerHTML" [@@bs.set]
end               
              
module MouseEvent = struct
  type t
  external addEventListener : string -> (t -> unit) -> unit = "" [@@bs.send.pipe: Mouse.t]
  external addClickEventListener : (_ [@bs.as "click"]) -> (t -> unit) -> unit = "addEventListener" [@@bs.send.pipe : Mouse.t]
end              

module Canvas = struct
  type t
  external width : t -> int = "" [@@bs.get]
  external height : t -> int = "" [@@bs.get]
  external getContext : string -> Context.t = "" [@@bs.send.pipe: t]
  external getElementById : string -> t option = "document.getElementById" [@@bs.val] [@@bs.return null_to_opt]

  external offsetLeft : t -> int = "" [@@bs.get]
  external offsetTop : t -> int = "" [@@bs.get]
  external scrollLeft : t -> int = "" [@@bs.get]
  external scrollTop : t -> int = "" [@@bs.get]
  external offsetParent : t -> Element.t option = "" [@@bs.get] [@@bs.return null_to_opt]
end

module CanvasEvent = struct
  type t
  external clientX : t -> int = "" [@@bs.get]
  external clientY : t -> int = "" [@@bs.get]

  external pageX : t -> int = "" [@@bs.get]
  external pageY : t -> int = "" [@@bs.get]                              
  external addEventListener : string -> (t -> unit) -> unit = "" [@@bs.send.pipe: Canvas.t]
  external addClickEventListener : (_ [@bs.as "click"]) -> (t -> unit) -> unit = "addEventListener" [@@bs.send.pipe : Canvas.t]
end

module Option = struct
  type t
  external value : t -> string = "" [@@bs.get]
end
                   
module Select = struct
  type t
  external getElementById : string -> t option = "document.getElementById" [@@bs.val] [@@bs.return null_to_opt]
  external selectedIndex : t -> int = "" [@@bs.get]
  external options : t -> Option.t array = "" [@@bs.get]
end               

module SelectEvent = struct
  type t
  external addEventListener : string -> (t -> unit) -> unit = "" [@@bs.send.pipe: Select.t]
  external addChangeEventListener : (_ [@bs.as "change"]) -> (t -> unit) -> unit = "addEventListener" [@@bs.send.pipe : Select.t]
end              

let isNone o =
  match o with
  | Some _o -> false
  | None -> true       

let isSome o =
  match o with
  | Some _o -> true
  | None -> false    

let relMouseCoords (c: Canvas.t) (e: CanvasEvent.t) =
  let totalOffsetX = ref @@ (Canvas.offsetLeft c) - (Canvas.scrollLeft c) in
  let totalOffsetY = ref @@ (Canvas.offsetTop c) - (Canvas.scrollTop c) in
  let currentElement = ref (Canvas.offsetParent c) in

  while isSome !currentElement do
    match !currentElement with
    | None -> ()
    | Some ec ->
       totalOffsetX := !totalOffsetX + (Element.offsetLeft ec) - (Element.scrollLeft ec);
       totalOffsetY := !totalOffsetY + (Element.offsetTop ec) - (Element.scrollTop ec);
       currentElement := Element.offsetParent ec
  done;

  let canvasX = (CanvasEvent.pageX e) - !totalOffsetX in
  let canvasY = (CanvasEvent.pageY e) - !totalOffsetY in

  (canvasX, canvasY)
  (* while  *)
(*
function relMouseCoords(event){
    var totalOffsetX = 0;
    var totalOffsetY = 0;
    var canvasX = 0;
    var canvasY = 0;
    var currentElement = this;

    do{
        totalOffsetX += currentElement.offsetLeft - currentElement.scrollLeft;
        totalOffsetY += currentElement.offsetTop - currentElement.scrollTop;
    }
    while(currentElement = currentElement.offsetParent)

    canvasX = event.pageX - totalOffsetX;
    canvasY = event.pageY - totalOffsetY;

    return {x:canvasX, y:canvasY}
}
         *)                  

type speed =
  | VerySlow
  | Slow
  | Normal
  | Fast
  | VeryFast

let parse_speed v =
  match v with
  | "very-slow" -> Some VerySlow
  | "slow" -> Some Slow
  | "normal" -> Some Normal
  | "fast" -> Some Fast
  | "very-fast" -> Some VeryFast
  | _ -> None

let speed_to_int s =
  match s with
  | VerySlow -> 1000
  | Slow -> 500
  | Normal -> 100
  | Fast -> 50
  | VeryFast -> 0
                 
type cell =
  | Dead
  | Alive

let size = 8
let foregroundColor = "#F36"
let backgroundColor = "#FFFFFF"
let pieceColor = "#FFFFFF"                    
let gridColor = "#C0C0C0"
let strokeWidth = 2
let cellSize = 18

type state =
  { run     : bool
  ; grid    : (cell array) array
  ; rows    : int
  ; columns : int
  ; speed   : speed
  }

let state: state ref =
  ref
    { run = true
    ; grid = [||]
    ; rows = 0
    ; columns = 0
    ; speed = Normal
    }
                     
let clear canvas context = (
  Context.setFillStyle context backgroundColor;
  context |> Context.fillRect 0 0 (canvas |> Canvas.width) (canvas |> Canvas.height);
)

let drawGrid canvas rows columns =
  let context = canvas |> Canvas.getContext "2d" in
  context |> clear canvas;
  Context.setLineWidth context (float_of_int strokeWidth);
  Context.setStrokeStyle context gridColor;
  context |> Context.beginPath;
  let maxHeight = rows * (strokeWidth + cellSize) in
  let maxWidth = columns * (strokeWidth + cellSize) in
  let width = ref strokeWidth in
  while !width < maxWidth do
    context |> Context.moveTo (float_of_int !width) (float_of_int 0);
    context |> Context.lineTo (float_of_int !width) (float_of_int (maxHeight + strokeWidth));
    width := !width + strokeWidth + cellSize;
  done;
  context |> Context.moveTo (float_of_int !width) (float_of_int 0);
  context |> Context.lineTo (float_of_int !width) (float_of_int (maxHeight + strokeWidth));
  let height = ref (strokeWidth - 1) in
  while !height < maxHeight do
    context |> Context.moveTo (float_of_int strokeWidth) (float_of_int !height);
    context |> Context.lineTo (float_of_int (maxWidth + strokeWidth)) (float_of_int !height);
    height := !height + strokeWidth + cellSize;
  done;
  context |> Context.moveTo (float_of_int strokeWidth) (float_of_int !height);
  context |> Context.lineTo (float_of_int (maxWidth + strokeWidth)) (float_of_int !height);
  context |> Context.stroke;
  Context.setFillStyle context foregroundColor

let calcXOffset x =
  (strokeWidth + 1) + (x * (strokeWidth + cellSize))

let calcYOffset y =
  ((strokeWidth / 2) + 1) + (y * (strokeWidth + cellSize))
  
let draw grid context = (
  Context.setFillStyle context foregroundColor;
  grid |> Array.iteri @@ fun x cells ->
  cells |> Array.iteri @@ fun y cell -> 
    match cell with 
    | Alive -> context |> Context.fillRect (calcXOffset x) (calcYOffset y) cellSize cellSize;
    | _ -> ();
)
                      
let calcIndex (x: int) (y: int) rows columns =
  let o_xIndex = ref None in
  let xCount = ref 0 in
  let maxWidth = columns * (strokeWidth + cellSize) in
  let width = ref strokeWidth in
  while !width < maxWidth && isNone !o_xIndex do
    if x >= !width && x <= !width + strokeWidth + cellSize
    then
      o_xIndex := Some !xCount
    else
      xCount := !xCount + 1;
      width := !width + strokeWidth + cellSize;
  done;
  match !o_xIndex with
  | None -> None
  | Some xIndex ->
     let o_yIndex = ref None in
     let yCount = ref 0 in
     let maxHeight = rows * (strokeWidth + cellSize) in     
     let height = ref (strokeWidth - 1) in
     while !height < maxHeight && isNone !o_yIndex do
       if y >= !height && y <= !height + strokeWidth + cellSize
       then
         o_yIndex := Some !yCount
       else
         yCount := !yCount + 1;
       height := !height + strokeWidth + cellSize;
     done;
     match !o_yIndex with
     | None -> None
     | Some yIndex -> Some (xIndex, yIndex)
         
let rows grid = Array.length grid

let columns grid =
  match (Array.length grid) with
  | length when 0 < length -> Array.length @@ Array.get grid 0
  | _ -> -1 
       
let index min max n =
  match n with
  | n when n < min -> max
  | n when max < n -> min
  | n -> n

let neighbours x y grid =
  [|(-1, -1); (-1, 0); (-1, 1); (0, -1); (0, 1); (1, -1); (1, 0); (1, 1)|]
  |> Array.map @@ fun (a, b) -> (
    let m = index 0 (rows grid - 1) (x + a) in
    let n = index 0 (columns grid - 1) (y + b) in
    Array.get (Array.get grid m) n
  )

let next grid =
  grid |> Array.mapi @@ fun x cells ->
    cells |> Array.mapi @@ fun y cell ->
      match (neighbours x y grid |> Js.Array.filter ((=) Alive) |> Array.length) with
      | 3 -> Alive
      | 2 -> cell
      | _ -> Dead

(* throw away result and return unit *)
let void _x = ()
           
let rec run canvas =
  let context = canvas |> Canvas.getContext "2d" in
  drawGrid canvas !state.rows !state.columns;
  context |> draw !state.grid;
  void @@ Js.Global.setTimeout (fun () ->
  if (!state.run)
  then
    begin
      state := {!state with grid = next !state.grid};
      Window.requestAnimationFrame @@ fun () -> run canvas
    end
  else ()
    ) @@ speed_to_int !state.speed

let step canvas =
  state := {!state with grid = next !state.grid; run = false};
  let context = canvas |> Canvas.getContext "2d" in
  drawGrid canvas !state.rows !state.columns;
  context |> draw !state.grid
  
let resetGrid rows columns =
  Random.self_init ();
  
  let grid =
    (Array.make_matrix rows columns () |> Array.map @@ fun cells ->
      cells |> Array.map @@ fun _ -> if (Random.bool ()) then Alive else Dead) in

  state := { !state with grid = grid }
  
(* draw each line, move*)
(* touchend *)
let main =
  let canvas = 
    match (Canvas.getElementById "canvas") with
    | None -> failwith "Cannot find the canvas"
    | Some canvas -> canvas in

  let play_button =
    match (Mouse.getElementById "play-button") with
    | None -> failwith "Cannot find the play-button"
    | Some play_button -> play_button in

  let step_button =
    match (Mouse.getElementById "step-button") with
    | None -> failwith "Cannot find the step-button"
    | Some step_button -> step_button in
  
  let random_reset_button =
    match (Mouse.getElementById "random-reset-button") with
    | None -> failwith "Cannot find the random-reset-button"
    | Some random_reset_button -> random_reset_button in

  let speed_select =
    match (Select.getElementById "speed-select") with
    | None -> failwith "Cannot find the speed-select"
    | Some speed_select -> speed_select in
  
  let handleClickGrid (e: CanvasEvent.t) = (
    let (x, y) = relMouseCoords canvas e in
    let o = calcIndex x y !state.rows !state.columns in
    match o with
    | None -> ()
    | Some (x, y) ->
       let g = !state.grid in
       let c = g.(x).(y) in
       let newC = match c with
         | Alive -> Dead
         | Dead -> Alive in
       g.(x).(y) <- newC;
       state := {!state with grid = g};
       let context = canvas |> Canvas.getContext "2d" in
       drawGrid canvas !state.rows !state.columns;
       context |> draw !state.grid;
    ) in
  canvas |> CanvasEvent.addClickEventListener (fun e -> handleClickGrid e);
  
  let toggleRun _unit =
    state := { !state with run = not !state.run };
    if !state.run
    then
      begin
        Mouse.setInnerHTML play_button "Pause";
        run canvas
      end
    else Mouse.setInnerHTML play_button "Play" in

  play_button |> MouseEvent.addClickEventListener (fun _unit -> toggleRun ());

  let handleSpeedSelectChange _event =(
    let selectedIndex = Select.selectedIndex speed_select in
    let option = (Select.options speed_select).(selectedIndex) in
    match parse_speed @@ Option.value option with
    | None -> () 
    | Some speed -> state := {!state with speed = speed}) in

  speed_select |> SelectEvent.addChangeEventListener handleSpeedSelectChange;

  step_button |> MouseEvent.addClickEventListener (fun _unit -> step canvas; Mouse.setInnerHTML play_button "Play";);

  let randomGrid _unit =
    resetGrid !state.rows !state.columns;
    state := { !state with run = false };
    run canvas in
  
  random_reset_button |> MouseEvent.addClickEventListener (fun _unit -> randomGrid (); Mouse.setInnerHTML play_button "Play";);
  
  let columns = ((canvas |> Canvas.width) - strokeWidth) / (strokeWidth + cellSize) in
  let rows = ((canvas |> Canvas.height) - strokeWidth) / (strokeWidth + cellSize) in
  
  Random.self_init ();
  
  let grid =
    (Array.make_matrix rows columns () |> Array.map @@ fun cells ->
      cells |> Array.map @@ fun _ -> if (Random.bool ()) then Alive else Dead) in

  state := { !state with rows = rows ; columns = columns; grid = grid };

  run canvas


lib/js/src/GameOfLife.js
'use strict';

var $$Array = require("bs-platform/lib/js/array.js");
var Random = require("bs-platform/lib/js/random.js");
var Caml_obj = require("bs-platform/lib/js/caml_obj.js");
var Caml_array = require("bs-platform/lib/js/caml_array.js");
var Caml_int32 = require("bs-platform/lib/js/caml_int32.js");
var Pervasives = require("bs-platform/lib/js/pervasives.js");
var Js_primitive = require("bs-platform/lib/js/js_primitive.js");

var Element = /* module */[];

var Context = /* module */[];

var Window = /* module */[];

var Mouse = /* module */[];

var MouseEvent = /* module */[];

var Canvas = /* module */[];

var CanvasEvent = /* module */[];

var Option = /* module */[];

var Select = /* module */[];

var SelectEvent = /* module */[];

function isNone(o) {
  if (o) {
    return false;
  } else {
    return true;
  }
}

function isSome(o) {
  if (o) {
    return true;
  } else {
    return false;
  }
}

function relMouseCoords(c, e) {
  var totalOffsetX = c.offsetLeft - c.scrollLeft | 0;
  var totalOffsetY = c.offsetTop - c.scrollTop | 0;
  var currentElement = Js_primitive.null_to_opt(c.offsetParent);
  while(isSome(currentElement)) {
    var match = currentElement;
    if (match) {
      var ec = match[0];
      totalOffsetX = (totalOffsetX + ec.offsetLeft | 0) - ec.scrollLeft | 0;
      totalOffsetY = (totalOffsetY + ec.offsetTop | 0) - ec.scrollTop | 0;
      currentElement = Js_primitive.null_to_opt(ec.offsetParent);
    }
    
  };
  var canvasX = e.pageX - totalOffsetX | 0;
  var canvasY = e.pageY - totalOffsetY | 0;
  return /* tuple */[
          canvasX,
          canvasY
        ];
}

function parse_speed(v) {
  switch (v) {
    case "fast" : 
        return /* Some */[/* Fast */3];
    case "normal" : 
        return /* Some */[/* Normal */2];
    case "slow" : 
        return /* Some */[/* Slow */1];
    case "very-fast" : 
        return /* Some */[/* VeryFast */4];
    case "very-slow" : 
        return /* Some */[/* VerySlow */0];
    default:
      return /* None */0;
  }
}

function speed_to_int(s) {
  switch (s) {
    case 0 : 
        return 1000;
    case 1 : 
        return 500;
    case 2 : 
        return 100;
    case 3 : 
        return 50;
    case 4 : 
        return 0;
    
  }
}

var foregroundColor = "#F36";

var backgroundColor = "#FFFFFF";

var gridColor = "#C0C0C0";

var state = [/* record */[
    /* run */true,
    /* grid : array */[],
    /* rows */0,
    /* columns */0,
    /* speed : Normal */2
  ]];

function clear(canvas, context) {
  context.fillStyle = backgroundColor;
  context.fillRect(0, 0, canvas.width, canvas.height);
  return /* () */0;
}

function drawGrid(canvas, rows, columns) {
  var context = canvas.getContext("2d");
  clear(canvas, context);
  context.lineWidth = 2;
  context.strokeStyle = gridColor;
  context.beginPath();
  var maxHeight = Caml_int32.imul(rows, 20);
  var maxWidth = Caml_int32.imul(columns, 20);
  var width = 2;
  while(width < maxWidth) {
    context.moveTo(width, 0);
    context.lineTo(width, maxHeight + 2 | 0);
    width = (width + 2 | 0) + 18 | 0;
  };
  context.moveTo(width, 0);
  context.lineTo(width, maxHeight + 2 | 0);
  var height = 1;
  while(height < maxHeight) {
    context.moveTo(2, height);
    context.lineTo(maxWidth + 2 | 0, height);
    height = (height + 2 | 0) + 18 | 0;
  };
  context.moveTo(2, height);
  context.lineTo(maxWidth + 2 | 0, height);
  context.stroke();
  context.fillStyle = foregroundColor;
  return /* () */0;
}

function calcXOffset(x) {
  return 3 + Caml_int32.imul(x, 20) | 0;
}

function calcYOffset(y) {
  return 2 + Caml_int32.imul(y, 20) | 0;
}

function draw(grid, context) {
  context.fillStyle = foregroundColor;
  return $$Array.iteri((function (x, cells) {
                return $$Array.iteri((function (y, cell) {
                              if (cell) {
                                context.fillRect(calcXOffset(x), calcYOffset(y), 18, 18);
                                return /* () */0;
                              } else {
                                return /* () */0;
                              }
                            }), cells);
              }), grid);
}

function calcIndex(x, y, rows, columns) {
  var o_xIndex = /* None */0;
  var xCount = 0;
  var maxWidth = Caml_int32.imul(columns, 20);
  var width = 2;
  while(width < maxWidth && isNone(o_xIndex)) {
    if (x >= width && x <= ((width + 2 | 0) + 18 | 0)) {
      o_xIndex = /* Some */[xCount];
    } else {
      xCount = xCount + 1 | 0;
    }
    width = (width + 2 | 0) + 18 | 0;
  };
  var match = o_xIndex;
  if (match) {
    var o_yIndex = /* None */0;
    var yCount = 0;
    var maxHeight = Caml_int32.imul(rows, 20);
    var height = 1;
    while(height < maxHeight && isNone(o_yIndex)) {
      if (y >= height && y <= ((height + 2 | 0) + 18 | 0)) {
        o_yIndex = /* Some */[yCount];
      } else {
        yCount = yCount + 1 | 0;
      }
      height = (height + 2 | 0) + 18 | 0;
    };
    var match$1 = o_yIndex;
    if (match$1) {
      return /* Some */[/* tuple */[
                match[0],
                match$1[0]
              ]];
    } else {
      return /* None */0;
    }
  } else {
    return /* None */0;
  }
}

function rows(grid) {
  return grid.length;
}

function columns(grid) {
  var length = grid.length;
  if (0 < length) {
    return Caml_array.caml_array_get(grid, 0).length;
  } else {
    return -1;
  }
}

function index(min, max, n) {
  if (Caml_obj.caml_lessthan(n, min)) {
    return max;
  } else if (Caml_obj.caml_lessthan(max, n)) {
    return min;
  } else {
    return n;
  }
}

function neighbours(x, y, grid) {
  return $$Array.map((function (param) {
                var m = index(0, grid.length - 1 | 0, x + param[0] | 0);
                var n = index(0, columns(grid) - 1 | 0, y + param[1] | 0);
                return Caml_array.caml_array_get(Caml_array.caml_array_get(grid, m), n);
              }), /* array */[
              /* tuple */[
                -1,
                -1
              ],
              /* tuple */[
                -1,
                0
              ],
              /* tuple */[
                -1,
                1
              ],
              /* tuple */[
                0,
                -1
              ],
              /* tuple */[
                0,
                1
              ],
              /* tuple */[
                1,
                -1
              ],
              /* tuple */[
                1,
                0
              ],
              /* tuple */[
                1,
                1
              ]
            ]);
}

function next(grid) {
  return $$Array.mapi((function (x, cells) {
                return $$Array.mapi((function (y, cell) {
                              var match = neighbours(x, y, grid).filter((function (param) {
                                      return Caml_obj.caml_equal(/* Alive */1, param);
                                    })).length;
                              if (match !== 2) {
                                if (match !== 3) {
                                  return /* Dead */0;
                                } else {
                                  return /* Alive */1;
                                }
                              } else {
                                return cell;
                              }
                            }), cells);
              }), grid);
}

function $$void() {
  return /* () */0;
}

function run(canvas) {
  var context = canvas.getContext("2d");
  drawGrid(canvas, state[0][/* rows */2], state[0][/* columns */3]);
  draw(state[0][/* grid */1], context);
  setTimeout((function () {
          if (state[0][/* run */0]) {
            var init = state[0];
            state[0] = /* record */[
              /* run */init[/* run */0],
              /* grid */next(state[0][/* grid */1]),
              /* rows */init[/* rows */2],
              /* columns */init[/* columns */3],
              /* speed */init[/* speed */4]
            ];
            requestAnimationFrame((function () {
                    return run(canvas);
                  }));
            return /* () */0;
          } else {
            return /* () */0;
          }
        }), speed_to_int(state[0][/* speed */4]));
  return /* () */0;
}

function step(canvas) {
  var init = state[0];
  state[0] = /* record */[
    /* run */false,
    /* grid */next(state[0][/* grid */1]),
    /* rows */init[/* rows */2],
    /* columns */init[/* columns */3],
    /* speed */init[/* speed */4]
  ];
  var context = canvas.getContext("2d");
  drawGrid(canvas, state[0][/* rows */2], state[0][/* columns */3]);
  return draw(state[0][/* grid */1], context);
}

function resetGrid(rows, columns) {
  Random.self_init(/* () */0);
  var grid = $$Array.map((function (cells) {
          return $$Array.map((function () {
                        if (Random.bool(/* () */0)) {
                          return /* Alive */1;
                        } else {
                          return /* Dead */0;
                        }
                      }), cells);
        }), $$Array.make_matrix(rows, columns, /* () */0));
  var init = state[0];
  state[0] = /* record */[
    /* run */init[/* run */0],
    /* grid */grid,
    /* rows */init[/* rows */2],
    /* columns */init[/* columns */3],
    /* speed */init[/* speed */4]
  ];
  return /* () */0;
}

var match = document.getElementById("canvas");

var canvas = match !== null ? match : Pervasives.failwith("Cannot find the canvas");

var match$1 = document.getElementById("play-button");

var play_button = match$1 !== null ? match$1 : Pervasives.failwith("Cannot find the play-button");

var match$2 = document.getElementById("step-button");

var step_button = match$2 !== null ? match$2 : Pervasives.failwith("Cannot find the step-button");

var match$3 = document.getElementById("random-reset-button");

var random_reset_button = match$3 !== null ? match$3 : Pervasives.failwith("Cannot find the random-reset-button");

var match$4 = document.getElementById("speed-select");

var speed_select = match$4 !== null ? match$4 : Pervasives.failwith("Cannot find the speed-select");

canvas.addEventListener("click", (function (e) {
        var e$1 = e;
        var match = relMouseCoords(canvas, e$1);
        var o = calcIndex(match[0], match[1], state[0][/* rows */2], state[0][/* columns */3]);
        if (o) {
          var match$1 = o[0];
          var y = match$1[1];
          var x = match$1[0];
          var g = state[0][/* grid */1];
          var c = Caml_array.caml_array_get(Caml_array.caml_array_get(g, x), y);
          var newC = c ? /* Dead */0 : /* Alive */1;
          Caml_array.caml_array_set(Caml_array.caml_array_get(g, x), y, newC);
          var init = state[0];
          state[0] = /* record */[
            /* run */init[/* run */0],
            /* grid */g,
            /* rows */init[/* rows */2],
            /* columns */init[/* columns */3],
            /* speed */init[/* speed */4]
          ];
          var context = canvas.getContext("2d");
          drawGrid(canvas, state[0][/* rows */2], state[0][/* columns */3]);
          return draw(state[0][/* grid */1], context);
        } else {
          return /* () */0;
        }
      }));

play_button.addEventListener("click", (function () {
        var init = state[0];
        state[0] = /* record */[
          /* run */!state[0][/* run */0],
          /* grid */init[/* grid */1],
          /* rows */init[/* rows */2],
          /* columns */init[/* columns */3],
          /* speed */init[/* speed */4]
        ];
        if (state[0][/* run */0]) {
          play_button.innerHTML = "Pause";
          return run(canvas);
        } else {
          play_button.innerHTML = "Play";
          return /* () */0;
        }
      }));

function handleSpeedSelectChange() {
  var selectedIndex = speed_select.selectedIndex;
  var option = Caml_array.caml_array_get(speed_select.options, selectedIndex);
  var match = parse_speed(option.value);
  if (match) {
    var init = state[0];
    state[0] = /* record */[
      /* run */init[/* run */0],
      /* grid */init[/* grid */1],
      /* rows */init[/* rows */2],
      /* columns */init[/* columns */3],
      /* speed */match[0]
    ];
    return /* () */0;
  } else {
    return /* () */0;
  }
}

speed_select.addEventListener("change", handleSpeedSelectChange);

step_button.addEventListener("click", (function () {
        step(canvas);
        play_button.innerHTML = "Play";
        return /* () */0;
      }));

function randomGrid() {
  resetGrid(state[0][/* rows */2], state[0][/* columns */3]);
  var init = state[0];
  state[0] = /* record */[
    /* run */false,
    /* grid */init[/* grid */1],
    /* rows */init[/* rows */2],
    /* columns */init[/* columns */3],
    /* speed */init[/* speed */4]
  ];
  return run(canvas);
}

random_reset_button.addEventListener("click", (function () {
        randomGrid(/* () */0);
        play_button.innerHTML = "Play";
        return /* () */0;
      }));

var columns$1 = (canvas.width - 2 | 0) / 20 | 0;

var rows$1 = (canvas.height - 2 | 0) / 20 | 0;

Random.self_init(/* () */0);

var grid = $$Array.map((function (cells) {
        return $$Array.map((function () {
                      if (Random.bool(/* () */0)) {
                        return /* Alive */1;
                      } else {
                        return /* Dead */0;
                      }
                    }), cells);
      }), $$Array.make_matrix(rows$1, columns$1, /* () */0));

var init = state[0];

state[0] = /* record */[
  /* run */init[/* run */0],
  /* grid */grid,
  /* rows */rows$1,
  /* columns */columns$1,
  /* speed */init[/* speed */4]
];

var main = run(canvas);

var size = 8;

var pieceColor = "#FFFFFF";

var strokeWidth = 2;

var cellSize = 18;

exports.Element = Element;
exports.Context = Context;
exports.Window = Window;
exports.Mouse = Mouse;
exports.MouseEvent = MouseEvent;
exports.Canvas = Canvas;
exports.CanvasEvent = CanvasEvent;
exports.Option = Option;
exports.Select = Select;
exports.SelectEvent = SelectEvent;
exports.isNone = isNone;
exports.isSome = isSome;
exports.relMouseCoords = relMouseCoords;
exports.parse_speed = parse_speed;
exports.speed_to_int = speed_to_int;
exports.size = size;
exports.foregroundColor = foregroundColor;
exports.backgroundColor = backgroundColor;
exports.pieceColor = pieceColor;
exports.gridColor = gridColor;
exports.strokeWidth = strokeWidth;
exports.cellSize = cellSize;
exports.state = state;
exports.clear = clear;
exports.drawGrid = drawGrid;
exports.calcXOffset = calcXOffset;
exports.calcYOffset = calcYOffset;
exports.draw = draw;
exports.calcIndex = calcIndex;
exports.rows = rows;
exports.columns = columns;
exports.index = index;
exports.neighbours = neighbours;
exports.next = next;
exports.$$void = $$void;
exports.run = run;
exports.step = step;
exports.resetGrid = resetGrid;
exports.main = main;
/* match Not a pure module */
